signal_scanner_bot/twitter.py

```diff
@@ -35,16 +36,73 @@ def get_api():
 ################################################################################
 def send_tweet(tweet: str, timestamp: datetime, api: tweepy.API) -> None:
     hashtags = " ".join(SEND_HASHTAGS)
-    if len(tweet + hashtags) >= 260:
-        # TODO: better
-        log.warning(f"Cannot tweet message, exceeds length: {tweet}")
-        return
-
-    formatted = dedent(
-        f"""
-    {tweet} @ {timestamp.strftime('%l:%M:%S%p').strip()}
-
-    {hashtags}
-    """
-    )
-    api.update_status(formatted)
+
+    # Check if tweet is longer than 280 minus a 20 char padding, add to list if
+    # not, continue with dividing up text if so
+    if len(tweet + hashtags) >= TWEET_MAX_SIZE - TWEET_PADDING:
+
+        # Split tweet into word list, create empty list to store serialized tweets, index to track last position
+        tweet_word_list = tweet.split(" ")
+        tweets_list = []
+        j = 0
+
+        # For loop through word list and enumerate the index because we'll need
+        # it but don't actually care about the list value so send to null
+        for index, _ in enumerate(tweet_word_list):
+
+            # Check if it is the first tweet, which will contain hashtags and
+            # timestamp. If not first it will only contain the text plus ellipses
+            # and for the final one drop the ellipses
+            if len(tweets_list) == 0:
+                sub_tweet = " ".join(tweet_word_list[j:index]) + " ..." + hashtags
+            elif index < len(tweet_word_list):
+                sub_tweet = " ".join(tweet_word_list[j:index]) + " ..."
+            elif index == len(tweet_word_list):
+                sub_tweet = " ".join(tweet_word_list[j:index])
+
+            # When length of tweet reaches >260 chars save to list and set
+            # base index for next tweet
+            if len(sub_tweet) > TWEET_MAX_SIZE - TWEET_PADDING:
+                tweets_list.append(" ".join(tweet_word_list[j:index - 1]) + " ...")
+                j = index - 1
```

Comment:
Just so I understand, it looks like this section is iteratively building up the tweet (with any extra bits that are necessary depending on where it is in the thread) and then when the tweet exceeds the maximum, uses the last known working position to determine the tweet size?

Reply:
Yeah, the logic is a little wonky here honestly. I'm open to criticism and suggestions for improvement. 

I'm declaring the tweet_list as an empty array (will be updated to be an empty array of string for the final merge) and then storing then checking if that array is empty or not. If it is empty, that's the first tweet in the thread and will thus need the the hashtags in addition to the ellipses, if not and the index is less than the length of the total length of the array that will be an intermediary tweet and only need the ellipses, and finally when the index reaches the last point (ie. index equals length) then we are at the last tweet and thus don't need the ellipses anymore.

I realized this morning that I accidentally made it work out by not actually saving the last tweet, because the last tweet skips the:
if len(sub_tweet) > TWEET_MAX_SIZE - TWEET_PADDING:

Part, because in my testing I've just been using the same string, I think I'm missing a potential edge case where the last tweet I'm building would go over the max tweet length we allow on the last index, and then get into a weird state. I need to think about this and figure out how to improve this logic.